dali/kernels/imgproc/jpeg/chroma_subsample_gpu.cuh

```diff
@@ -0,0 +1,241 @@
+  T cr_val = cr(chroma_x, chroma_y);
+
+  write_vec(&out(x, y),
+            ycbcr_to_rgb(vec<3, T>(luma(luma_x, luma_y), cb_val, cr_val)));
+  if (horz_subsample && vert_subsample) {
+    write_vec(&out(x + 1, y),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x + 1, luma_y), cb_val, cr_val)));
+    write_vec(&out(x, y + 1),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x, luma_y + 1), cb_val, cr_val)));
+    write_vec(&out(x + 1, y + 1),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x + 1, luma_y + 1), cb_val, cr_val)));
+  } else if (horz_subsample) {
+    write_vec(&out(x + 1, y),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x + 1, luma_y), cb_val, cr_val)));
+  } else if (vert_subsample) {
+    write_vec(&out(x, y + 1),
+              ycbcr_to_rgb(vec<3, T>(luma(luma_x, luma_y + 1), cb_val, cr_val)));
+  }
+}
+
+
+template <bool horz_subsample, bool vert_subsample>
+__global__ void ChromaSubsampleDistortion(const SampleDesc *samples,
+                                          const kernels::BlockDesc<2> *blocks) {
+  using T = uint8_t;
+  constexpr int luma_blk_h = 8 << vert_subsample;
+  constexpr int luma_blk_w = 8 << horz_subsample;
+  constexpr ivec<2> luma_blk_sz{luma_blk_w, luma_blk_h};
+  constexpr i64vec<2> luma_blk_strides{1, luma_blk_w};
+  constexpr ivec<2> chroma_blk_sz{8, 8};
+  constexpr i64vec<2> chroma_blk_strides{1, 8};
+
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  int y_start = threadIdx.y + block.start.y;
+  int x_start = threadIdx.x + block.start.x;
+  if (y_start >= block.end.y || x_start >= block.end.x) {
+    return;
+  }
+
+  // Assuming CUDA block has:
+  // - width 32, leads to 4 horizontal blocks of 8
+  // - height 8, so a single block 8x8 fits vertically
+  __shared__ T luma_blk[4][luma_blk_h][luma_blk_w];
+  __shared__ T cb_blk[4][8][8];
+  __shared__ T cr_blk[4][8][8];
+
+  int blk_idx = threadIdx.x / 8;
+  int local_x = threadIdx.x % 8;
+  int local_y = threadIdx.y;
+
+  const Surface2D<const uint8_t> in = {
+    sample.in, sample.size, 3, sample.strides, 1
+  };
+
+  const Surface2D<uint8_t> out = {
+    sample.out, sample.size, 3, sample.strides, 1
+  };
+
+  const Surface2D<T> luma = {
+    &luma_blk[blk_idx][0][0], luma_blk_sz, 1, luma_blk_strides, 1
+  };
+
+  const Surface2D<T> cb = {
+    &cb_blk[blk_idx][0][0], chroma_blk_sz, 1, chroma_blk_strides, 1
+  };
+
+  const Surface2D<T> cr = {
+    &cr_blk[blk_idx][0][0], chroma_blk_sz, 1, chroma_blk_strides, 1
+  };
+
+  for (int chroma_y = y_start; chroma_y < block.end.y; chroma_y += blockDim.y) {
+    for (int chroma_x = x_start; chroma_x < block.end.x; chroma_x += blockDim.x) {
+      int y = chroma_y << vert_subsample;
+      int x = chroma_x << horz_subsample;
+
+      ivec2 blk_offset{local_x, local_y};
+      ivec2 offset{x, y};
+
+      rgb_to_ycbcr_chroma_subsample<horz_subsample, vert_subsample>(
+          blk_offset, offset, luma, cb, cr, in);
+
+      __syncthreads();
```

Comment:
In _this_ kernel, the synchronization is not necessary - you're accessing only the elements produced by this thread.

Reply:
I am aware of that. I just put it here, because I was planning to build on top of that. I also don't need the shared memory.